src/core/leader_schedule.zig

```diff
@@ -9,205 +9,318 @@ const Pubkey = sig.core.Pubkey;
-
-    _, const slot_index = bank_fields.epoch_schedule.getEpochAndSlotIndex(bank_fields.slot);
-    const epoch_start_slot = bank_fields.slot - slot_index;
-    return SingleEpochLeaderSchedule{
-        .allocator = allocator,
-        .slot_leaders = slot_leaders,
-        .start_slot = epoch_start_slot,
-    };
-}
+    pub fn getSlotLeaderNoCompute(self: *LeaderScheduleCache, slot: Slot) ?Pubkey {
+        const epoch, const slot_index = self.epoch_schedule.getEpochAndSlotIndex(slot);
+        return self.getSlotLeader(epoch, slot_index);
+    }
+
+    pub fn getSlotLeaderMaybeCompute(self: *LeaderScheduleCache, slot: Slot, bank: *const BankFields) !sig.core.Pubkey {
+        const epoch, const slot_index = self.epoch_schedule.getEpochAndSlotIndex(slot);
+        if (self.getSlotLeader(epoch, slot_index)) |leader| return leader;
 
-pub fn leaderScheduleFromRpc(allocator: Allocator, start_slot: Slot, rpc_client: *RpcClient) !SingleEpochLeaderSchedule {
-    const rpc_leader_schedule_response = try rpc_client.getLeaderSchedule(allocator, null, .{});
-    defer rpc_leader_schedule_response.deinit();
-    const rpc_leader_schedule = try rpc_leader_schedule_response.result();
+        const leader_schedules, var leader_schedule_lg = self.leader_schedules_rw.writeWithLock();
+        defer leader_schedule_lg.unlock();
 
-    var num_leaders: u64 = 0;
-    for (rpc_leader_schedule.values()) |leader_slots| {
-        num_leaders += leader_slots.len;
+        const entry = try leader_schedules.getOrPut(epoch);
+        if (!entry.found_existing) {
+            if (leader_schedules.count() == MAX_CACHED_LEADER_SCHEDULES) {
+                _ = leader_schedules.swapRemove(std.mem.min(Epoch, leader_schedules.keys()));
+            }
+            entry.value_ptr.* = try LeaderSchedule.fromBank(
+                self.allocator,
+                epoch,
+                bank,
+            );
+        }
+        return entry.value_ptr.slot_leaders[slot_index];
     }
 
-    const Record = struct { slot: Slot, key: Pubkey };
+    pub fn getSlotLeaderMaybeComputeRpc(self: *LeaderScheduleCache, slot: Slot, rpc_client: *RpcClient) !sig.core.Pubkey {
+        const epoch, const slot_index = self.epoch_schedule.getEpochAndSlotIndex(slot);
+        if (self.getSlotLeader(epoch, slot_index)) |leader| return leader;
 
-    var leaders_index: usize = 0;
-    var leaders = try allocator.alloc(Record, num_leaders);
-    defer allocator.free(leaders);
+        const leader_schedules, var leader_schedule_lg = self.leader_schedules_rw.writeWithLock();
+        defer leader_schedule_lg.unlock();
 
-    var rpc_leader_iter = rpc_leader_schedule.iterator();
-    while (rpc_leader_iter.next()) |entry| {
-        const key = try Pubkey.fromString(entry.key_ptr.*);
-        for (entry.value_ptr.*) |slot| {
-            leaders[leaders_index] = .{ .slot = slot, .key = key };
-            leaders_index += 1;
+        const entry = try leader_schedules.getOrPut(epoch);
```

Comment:
maybe want to have an `insertLeaderSchedule` method call here

Reply:
265ed12